### DISCUSSION TITLE
working with branches: working in build branch, cleaning up code in main branch

### DISCUSSION THREAD

**Discussion by Hifilo** [community/community#57961] [New to GitHub] [answered]
I recently decided to branch my project to have a repo where I can do all my changes, comment, trial/error, A/B etc. and when after a while I would upload the changes to the main branch, where I would clean up all the comments/spaghetti. 

`Main <- Build `

However this seemed to cause some problems. When I cleaned up the comments in main, the build branch became out of sync. Eventually, it appeared that I may have fallen into some merge conflict/rebase loop going back & forth re-merging into one another. I eventually did a final pull/merge into main & deleted the build branch.

The idea is that the `build` branch I'm working in would never pull unless I explicitly checkout & pull from the `main`, while still being able to push the changes I make in `build` to `main`

If I wanted to achieve this do I require another branch between main & build where I would do the cleanup `Main <- cleanup <- Build `, or am I going about this incorrectly?

**Comment by airtower-luna** [comment]
Do I understand right that `build` is supposed to be a stable branch, that only gets changes from main when they're final? If yes, you don't need another branch.

> However this seemed to cause some problems. When I cleaned up the comments in main, the build branch became out of sync.

What kind of cleanup did you do there? If it involves rebasing, any rebasing of commits that are already in `build` *will* make a mess once you try to merge: `build` still has the old history, and then you're merging in the new commits, too.

If you want to have a stable branch that only ever pulls from the one where you do development, don't pull until whatever history you're pulling is absolutely final. Otherwise you either have to live with the weird mix, or end up having to reset the stable branch now and then.

**Reply by Hifilo** [reply]
> Do I understand right that `build` is supposed to be a stable branch, that only gets changes from main when they're final? If yes, you don't need another branch.

the opposite.  `main` is the stable branch that only gets changes from build when they're final.
In `main` I wouldn't change any actual code, I would remove comments inside blocks that disabled certain lines of code. 

for example in the `style.css` file in `build` I would have

```css
.body {
    box-sizing: border-box;
    /* box-shadow: var( --box-shadow-2 ); */
    /* max-width: 100vw; */ 
    width:100vw;
    margin: 1em;
    /* margin: 2em; */
}
```
and in `main` the same `style.css` file would be cleaned up to:

```css
.body {
    box-sizing: border-box;
    width:100vw;
    margin: 1em;
}
```

I would still have all the disabled code inside `build`, while the stable `main` was cleaned up.

But based on how you've laid it out, git doesn't work that way

**Reply by airtower-luna** [reply, marked as answer]
It's not technically impossible, but generally not advisable, and trying to make it work would be a lot of work (deleting duplicate commits, rebasing, etc.). :sweat_smile: 

The common approach is to clean up code *before* merging to the stable branch, and consider what's been merged final, immutable history (except maybe in exceptional circumstances). Also keep in mind that any rebasing on a stable branch *will* cause trouble as soon as you collaborate with others.

I'd recommend working with *topic branches*: Start a branch for a new thing (bugfix, feature, whatever) that you're working on. While on that branch you can rebase, clean up, etc. as needed without any trouble. And once you're happy with the work, merge. See: [Git Branching - Branching Workflows](https://git-scm.com/book/en/v2/Git-Branching-Branching-Workflows) :slightly_smiling_face: 

In most cases I'd also recommend against committing commented-out code in the first place, if necessary use `git add -p` to commit only the changes within a file that you actually want to keep. If those uncommitted changes get in the way (e.g. when changing/merging branches) you can always [stash them](https://git-scm.com/book/en/v2/Git-Tools-Stashing-and-Cleaning).

**Reply by Hifilo** [reply]
Okay. I assumed what I was trying to do was possible. I guess I could make a copy and stash it, or add it to `.gitignore`. 

I appreciate you providing the thorough breakdown. Thank you.
